File: 3804-MaximizeActiveSectionWithTradeIi/3804-MaximizeActiveSectionWithTradeIi.cpp
// Last updated: 12/25/2025, 7:12:00 PM
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int left = 0, right = 0, maxVal = 0;
    Node() {}
    Node(int l, int r, int m) : left(l), right(r), maxVal(m) {}
};

int cmax(int x, int y) { return x > y ? x : y; }
int cmin(int x, int y) { return x < y ? x : y; }

class Solution {
public:
    vector<int> maxActiveSectionsAfterTrade(string s, vector<vector<int>>& queries) {
        int n = s.size();
        int total_ones = count(s.begin(), s.end(), '1');
        int k = floor(log2(n)) + 1;
        vector<vector<Node>> s_table(k, vector<Node>(n));

        vector<int> zero_prefix(n + 1, 0);
        for (int i = 0; i < n; i++) {
            zero_prefix[i + 1] = zero_prefix[i] + (s[i] == '0');
        }

        vector<int> zeros_left(n, -1), zeros_right(n, -1);
        char prev = s[0];
        int index = -1;
        for (int i = 1; i < n; i++) {
            if (prev != s[i]) {
                if (prev == '0') index = i - 1;
                prev = s[i];
            }
            zeros_left[i] = index;
        }

        prev = s[n - 1];
        index = -1;
        for (int i = n - 2; i >= 0; i--) {
            if (prev != s[i]) {
                if (prev == '0') index = i + 1;
                prev = s[i];
            }
            zeros_right[i] = index;
        }

        vector<int> zero_prefix_left(n, 0), zero_prefix_right(n, n - 1);
        prev = s[0];
        int prev_index = 0;
        for (int i = 1; i < n; i++) {
            if (prev != s[i]) {
                prev = s[i];
                prev_index = i;
            }
            if (s[i] == '0') zero_prefix_left[i] = prev_index;
        }

        prev = s[n - 1];
        prev_index = n - 1;
        for (int i = n - 2; i >= 0; i--) {
            if (prev != s[i]) {
                prev = s[i];
                prev_index = i;
            }
            if (s[i] == '0') zero_prefix_right[i] = prev_index;
        }

        for (int i = 0; i < n; i++) {
            if (s[i] == '0') {
                s_table[0][i].left = 0;
                s_table[0][i].right = 0;
            } else {
                s_table[0][i].left = 1;
                s_table[0][i].right = 1;
            }
        }

        auto getMaxNode = [&](int l, int mid, int r, Node& left, Node& right) {
            Node node;
            node.left = left.left;
            node.right = right.right;

            int curr = cmax(left.maxVal, right.maxVal);

            if (left.right == 0) {
                if (right.left == 0) {
                    int left_index = zeros_left[mid];
                    if (left_index != -1 && left_index >= l) {
                        left_index = cmax(l, zero_prefix_left[left_index]);
                        int right_index = cmin(r, zero_prefix_right[mid + 1]);
                        curr = cmax(curr, zero_prefix[right_index + 1] - zero_prefix[left_index]);
                    }
                }
                int left_index = cmax(l, zero_prefix_left[mid]);
                int right_index = zeros_right[mid + 1];
                if (right_index != -1 && right_index <= r) {
                    right_index = cmin(r, zero_prefix_right[right_index]);
                    curr = cmax(curr, zero_prefix[right_index + 1] - zero_prefix[left_index]);
                }
            } else {
                int left_index = zeros_left[mid];
                if (left_index != -1 && left_index >= l) {
                    left_index = cmax(l, zero_prefix_left[left_index]);
                    if (right.left == 1) {
                        int right_index = zeros_right[mid + 1];
                        if (right_index != -1 && right_index <= r) {
                            right_index = cmin(r, zero_prefix_right[right_index]);
                            curr = cmax(curr, zero_prefix[right_index + 1] - zero_prefix[left_index]);
                        }
                    } else {
                        int right_index = cmin(r, zero_prefix_right[mid + 1]);
                        curr = cmax(curr, zero_prefix[right_index + 1] - zero_prefix[left_index]);
                    }
                }
            }

            node.maxVal = curr;
            return node;
        };

        for (int power = 1; power < k; power++) {
            int m = 1 << power;
            int prev_m = m >> 1;
            for (int i = 0; i + m <= n; i++) {
                s_table[power][i] = getMaxNode(i, i + prev_m - 1, i + m - 1, s_table[power - 1][i], s_table[power - 1][i + prev_m]);
            }
        }

        vector<int> ret;
        for (auto& q : queries) {
            int l = q[0], r = q[1];
            int dis = r - l + 1;
            int power = 0;
            Node* prev = nullptr;
            int x = l;
            while (dis) {
                if (dis & 1) {
                    Node curr = s_table[power][l];
                    int mid = l - 1;
                    l += 1 << power;
                    if (prev != nullptr) {
                        Node combined = getMaxNode(x, mid, l - 1, *prev, curr);
                        *prev = combined;
                    } else {
                        prev = new Node(curr);
                    }
                }
                dis >>= 1;
                power++;
            }
            ret.push_back(prev->maxVal + total_ones);
            delete prev;
        }
        return ret;
    }
};

File: 3804-MaximizeActiveSectionWithTradeIi/3804-MaximizeActiveSectionWithTradeIi.py
# Last updated: 12/25/2025, 7:12:05 PM
class Solution:
    def maxActiveSectionsAfterTrade(self, s, queries):
        n = len(s)
        active_count = s.count('1')
        
        segments = []
        start = 0
        for i in range(n):
            if i == n - 1 or s[i] != s[i + 1]:
                segments.append((start, i - start + 1))
                start = i + 1
        segment_count = len(segments)
        
        max_power = 20
        INF = 10**9
        NEG_INF = -INF
        rmq = [[NEG_INF] * segment_count for _ in range(max_power)]
        
        for i in range(segment_count):
            if s[segments[i][0]] == '0' and i + 2 < segment_count:
                rmq[0][i] = segments[i][1] + segments[i + 2][1]
        
        for power in range(1, max_power):
            range_len = 1 << power
            for i in range(segment_count - range_len + 1):
                rmq[power][i] = max(rmq[power - 1][i],
                                    rmq[power - 1][i + (range_len >> 1)])
        
        def get_max_in_range(l, r):
            if l > r:
                return NEG_INF
            p = (r - l + 1).bit_length() - 1
            return max(rmq[p][l], rmq[p][r - (1 << p) + 1])
        
        result = []
        for left, right in queries:
            left_index = bisect_right(segments, (left, INF)) - 1
            right_index = bisect_right(segments, (right, INF)) - 1
            
            if right_index - left_index + 1 <= 2:
                result.append(active_count)
                continue
            
            def get_segment_size(i):
                if i == left_index:
                    return segments[left_index][1] - (left - segments[left_index][0])
                if i == right_index:
                    return right - segments[right_index][0] + 1
                return segments[i][1]
            
            def calculate_new_sections(i):
                if i < 0 or i + 2 >= segment_count or s[segments[i][0]] == '1':
                    return NEG_INF
                return get_segment_size(i) + get_segment_size(i + 2)
            
            best_increase = max(get_max_in_range(left_index + 1, right_index - 3), 0)
            best_increase = max(best_increase, calculate_new_sections(left_index))
            best_increase = max(best_increase, calculate_new_sections(right_index - 2))
            
            result.append(active_count + best_increase)
        
        return result